src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Patterns.cs

```diff
@@ -1,164 +1,581 @@
+            {
+                this._localRewriter = localRewriter;
+                this._factory = localRewriter._factory;
+                this._tempAllocator = new DagTempAllocator(localRewriter._factory);
+                this._loweredInput = loweredInput;
+                this._inputTemp = new BoundDagTemp(loweredInput.Syntax, loweredInput.Type, null, 0);
+                this._conjunctBuilder = ArrayBuilder<BoundExpression>.GetInstance();
+                this._sideEffectBuilder = ArrayBuilder<BoundExpression>.GetInstance();
             }
-        }
 
-        BoundPattern LowerPattern(BoundPattern pattern)
-        {
-            switch (pattern.Kind)
+            public void Dispose()
             {
-                case BoundKind.DeclarationPattern:
-                    {
-                        var declPattern = (BoundDeclarationPattern)pattern;
-                        return declPattern.Update(declPattern.Variable, VisitExpression(declPattern.VariableAccess), declPattern.DeclaredType, declPattern.IsVar);
-                    }
-                case BoundKind.ConstantPattern:
-                    {
-                        var constantPattern = (BoundConstantPattern)pattern;
-                        return constantPattern.Update(VisitExpression(constantPattern.Value), constantPattern.ConstantValue);
-                    }
-                default:
-                    return pattern;
+                _conjunctBuilder.Free();
+                _sideEffectBuilder.Free();
+                _tempAllocator.Dispose();
             }
-        }
 
-        private BoundExpression MakeIsConstantPattern(BoundConstantPattern loweredPattern, BoundExpression loweredInput)
-        {
-            return CompareWithConstant(loweredInput, loweredPattern.Value);
-        }
+            public class DagTempAllocator : IDisposable
+            {
+                private readonly SyntheticBoundNodeFactory _factory;
+                private readonly PooledDictionary<BoundDagTemp, BoundExpression> _map = PooledDictionary<BoundDagTemp, BoundExpression>.GetInstance();
+                private readonly ArrayBuilder<LocalSymbol> _temps = ArrayBuilder<LocalSymbol>.GetInstance();
 
-        private BoundExpression MakeIsDeclarationPattern(BoundDeclarationPattern loweredPattern, BoundExpression loweredInput)
-        {
-            Debug.Assert(((object)loweredPattern.Variable == null && loweredPattern.VariableAccess.Kind == BoundKind.DiscardExpression) ||
-                         loweredPattern.Variable.GetTypeOrReturnType() == loweredPattern.DeclaredType.Type);
+                public DagTempAllocator(SyntheticBoundNodeFactory factory)
+                {
+                    this._factory = factory;
+                }
 
-            if (loweredPattern.IsVar)
-            {
-                var result = _factory.Literal(true);
+                public void Dispose()
```

Comment:
Seems strange to suddenly switch conventions from explicitly calling `Free` to using `IDisposable` and `Dispose`.

Reply:
Agree; this should be Free(). Recorded in [https://github.com/dotnet/roslyn/issues/23222](https://github.com/dotnet/roslyn/issues/23222) .

---
In reply to: [155095606](https://github.com/dotnet/roslyn/pull/23209#discussion_r155095606) [](ancestors = 155095606)